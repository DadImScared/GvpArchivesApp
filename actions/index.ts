export { default as autoComplete } from "./autoComplete";
export { default as audioPlayer } from "./audioPlayer";
export { default as itemsByCategory } from "./itemsByCategory";
export { default as itemsById } from "./itemsById";
export { default as loading } from "./loading";
export { default as search } from "./search";
