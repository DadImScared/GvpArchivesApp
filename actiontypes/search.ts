export const UPDATE_SEARCH_QUERY = "search/UPDATE_SEARCH_QUERY";
export const TOGGLE_CATEGORIES = "search/TOGGLE_CATEGORIES";
export const SELECT_ALL_CATEGORIES = "search/SELECT_ALL_CATEGORIES";
export const REMOVE_ALL_CATEGORIES = "search/REMOVE_ALL_CATEGORIES";
