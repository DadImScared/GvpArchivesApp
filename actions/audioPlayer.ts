import { action } from "typesafe-actions";

import * as AudioActionTypes from "../actiontypes/audioPlayer";

export const togglePlaying = () => action(AudioActionTypes.TOGGLE_PLAYING);

export const setPlaying = (playing: boolean) => action(AudioActionTypes.SET_PLAYING, { playing });

export const showPlayer = (shouldShowPlayer: boolean) => {
  return action(AudioActionTypes.SHOW_PLAYER, { showPlayer: shouldShowPlayer });
};

export const setPlaylist = (name: string, items: string[], currentIndex = 0) => {
  return action(
    AudioActionTypes.SET_PLAYLIST,
    {
      currentIndex,
      items,
      name,
      playerType: "playlist" as any,
      playing: true,
      showPlayer: true,
    }
  );
};

export const updateIndex = (newIndex: number) => action(AudioActionTypes.UPDATE_CURRENT_INDEX, { newIndex });

export const setSong = (songUrl: string, songName: string) => {
  return action(
    AudioActionTypes.SET_SONG,
    {
      songName,
      songUrl,
      playerType: "song",
      playing: true,
      showPlayer: true
    }
  );
};

export interface IAudioPlayerActions {
  updateIndex: typeof updateIndex;
  togglePlaying: typeof togglePlaying;
  setSong: typeof setSong;
  setPlaying: typeof setPlaying;
  setPlaylist: typeof setPlaylist;
  showPlayer: typeof showPlayer;
}

export default {
  setPlaying,
  setPlaylist,
  setSong,
  showPlayer,
  togglePlaying,
  updateIndex,
};
