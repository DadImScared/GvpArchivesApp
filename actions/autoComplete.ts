import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/autoComplete";

const addAutoComplete = (query: string, categories: string[], results: string[]) => {
  return action(ActionTypes.ADD_AUTOCOMPLETE, { query, categories, results });
};

export default {
  addAutoComplete
};
