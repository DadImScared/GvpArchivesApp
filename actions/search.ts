import axios, { AxiosPromise } from "axios";
import { AnyAction, Dispatch } from "redux";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/search";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { getQueryId } from "../reducers/search";
import autoComplete from "./autoComplete";
import { BASE_URL } from "./itemsById";

const autoCompleteUrl = `${BASE_URL}/api/v1/completeme/`;

export const isValidValidTimeToLive = (time: number) => time ? time >= +new Date() : false;

const updateSearchQuery = (query: string) => action(ActionTypes.UPDATE_SEARCH_QUERY, query);

type ApiCallCallback<ReturnType> = (
  state: IReducerState,
  dispatch: Dispatch
) => ReturnType;

type IAfterApiCall<Data> = (
  data: Data,
  state: IReducerState,
  dispatch: Dispatch
) => Promise<AnyAction[]> | AnyAction[];

export interface IApiCallConfig<Data> {
  afterApiCall?: IAfterApiCall<Data>;
  apiCall: ApiCallCallback<AxiosPromise<any>>;
  beforeApiCall?: ApiCallCallback<Promise<AnyAction[]> | AnyAction[]>;
  loadingId?: ((state: IReducerState) => string) | string;
  shouldCallApi?: ApiCallCallback<boolean | AxiosPromise<void>>;
}

export const getQuerySuggestions = (query: string, categories: string[]): IApiCallConfig<string[]> => {
  const queryId = getQueryId(query, categories);
  const loadingId = `${AUTOCOMPLETE_QUERY}${queryId}`;
  return {
    loadingId,
    afterApiCall: (data) => {
      return [
        autoComplete.addAutoComplete(
          query,
          categories,
          data.map((textObj: any) => textObj.text)
        )
      ];
    },
    apiCall: () => axios.get(`${autoCompleteUrl}${query}`),
    shouldCallApi: (state) => {
      const isLoading = state.loading[loadingId];
      if (!query || isLoading) return false;

      const currentAutoComplete = state.autoComplete[queryId];
      if (!currentAutoComplete) return true;

      return isValidValidTimeToLive(currentAutoComplete.timeToLive) ?
        false
        :
        axios.get(`${autoCompleteUrl}${query}`);
    }
  };
};

export default {
  updateSearchQuery
};
