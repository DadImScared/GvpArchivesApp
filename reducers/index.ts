import _ from "lodash";
import { combineReducers, DeepPartial } from "redux";

import audioPlayer, { getInitialAudioPlayerState, IAudioPlayerState } from "./audioPlayer";
import autoComplete, { getInitialAutoCompleteState, IAutoCompleteState } from "./autoComplete";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";
import recentSearches, { RecentSearches } from "./recentSearches";
import search, { getInitialSearchState, ISearchState } from "./search";
import searchResults, { ISearchResults } from "./searchResults";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  autoComplete: IAutoCompleteState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
  recentSearches: RecentSearches;
  search: ISearchState;
  searchResults: ISearchResults;
}

export const getInitialReducerState = (injectedState: DeepPartial<IReducerState> = {}) => _.merge({
  audioPlayer: getInitialAudioPlayerState(),
  autoComplete: getInitialAutoCompleteState(),
  itemsByCategory: {},
  itemsById: {},
  loading: {},
  recentSearches: [],
  search: getInitialSearchState(),
  searchResults: {}
}, injectedState);

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  autoComplete,
  itemsByCategory,
  itemsById,
  loading,
  recentSearches,
  search,
  searchResults
});

export default reducer;
