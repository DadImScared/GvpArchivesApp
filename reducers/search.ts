import _ from "lodash";
import { ActionType } from "typesafe-actions";

import { search } from "../actions";
import * as ActionTypes from "../actiontypes/search";
import { IReducerState } from "./index";

export interface ISearchState {
  categories: string[];
  query: string;
}

export type SearchStateAndActions<
  T extends keyof ISearchState | keyof typeof search
  > = Pick<ISearchState & typeof search, T>;

export const CATEGORIES = [
  "bhagavatpatrika",
  "book",
  "harikatha",
  "harmonistmagazine",
  "harmonistmonthly",
  "lecture",
  "movie",
  "song"
];

export const getInitialSearchState = (): ISearchState => ({
  categories: CATEGORIES,
  query: ""
});

export default function(state = getInitialSearchState(), action: ActionType<typeof search>) {
  switch (action.type) {
    case ActionTypes.UPDATE_SEARCH_QUERY:
      return {
        ...state,
        query: action.payload
      };
    case ActionTypes.TOGGLE_CATEGORIES:
      const { payload } = action;
      const { categories } = state;
      return {
        ...state,
        categories: categories.includes(payload) ?
          categories.filter((category) => category !== action.payload)
          :
          [...categories, payload]
      };
    case ActionTypes.SELECT_ALL_CATEGORIES:
      return {
        ...state,
        categories: [...CATEGORIES]
      };
    case ActionTypes.REMOVE_ALL_CATEGORIES:
      return {
        ...state,
        categories: []
      };
    default:
      return state;
  }
}

export const getQueryId = (query: string, categories: string[]) => {
  return `${query}-${_.sortBy(categories, (o) => o).join(",")}`;
};

export const getSearch = (state: IReducerState) => state.search;
