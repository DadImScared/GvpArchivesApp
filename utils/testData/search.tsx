import * as React from "react";
import * as renderer from "react-test-renderer";

import { action } from "@storybook/addon-actions";
import _ from "lodash";

import { getQueryId } from "../../reducers/search";

export const makeTree = (Component: any, props: any = {}) => (injectedProps: any = {}) => {
  return renderer.create(<Component {...props} {...injectedProps} />);
};

const params = {
  categories: ["book", "song"],
  query: "radha",
  queryId: getQueryId("radha", ["book", "song"])
} as any;

export const getMockFunction = ({
  storybook = false,
  actionName = "",
  cb = false
}: any) => {
  if (storybook) {
    return (...args: any) => {
      action(actionName)(...args);
      if (cb) {
        return cb(...args);
      }
    };
  }
  return cb ? jest.fn(cb) : jest.fn();
};

export const searchTestProps = ({ props = {}, storybook = false } = {}) => (_.merge({
  addRecentSearch: getMockFunction({ storybook, actionName: "addRecentSearch" }),
  getSearchResults: getMockFunction({ storybook, actionName: "getSearchResults" }),
  isLoading: false,
  navigation: {
    getParam: getMockFunction({ storybook, actionName: "getParam", cb: (param: string) => params[param] }),
    push: getMockFunction({ storybook, actionName: "push" })
  },
  results: [
    {
      category: "book",
      item_id: "item-1",
      link: "item-1-link",
      title: "item-1-title"
    }
  ],
  suggestions: ["Suggestion-1", "Suggestion-2", "Suggestion-3"],
  timeToLive: 12345
}, props));
