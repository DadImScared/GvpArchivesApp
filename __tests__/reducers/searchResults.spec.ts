import moment from "moment";

import { searchResults } from "../../actions";
import { getQueryId } from "../../reducers/search";
import reducer from "../../reducers/searchResults";

describe("searchResults reducer", () => {
  it("should add search", () => {
    moment.now = () => 1558593861568;
    expect(
      reducer(
        {},
        searchResults.addSearch(
          getQueryId("radha", ["book", "movie"]),
          [{ item_id: "1", highlightedBody: "search body" }],
          [],
          "/nextpage"
        )
      )
    ).toMatchSnapshot();
    moment.now = () => +new Date();
  });

  it("should update search", () => {
    const queryId = getQueryId("radha", ["book", "movie"]);
    expect(
      reducer(
        {
          [queryId]: {
            nextPage: false,
            results: [{ item_id: "2", highlightedBody: "search body" }],
            suggestions: ["suggestion"],
            timeToLive: 1558593861568
          }
        },
        searchResults.updateSearch(
          queryId,
          [
            { item_id: "1", highlightedBody: "search body" },
            { item_id: "2", highlightedBody: "search body" }
          ],
          [],
          "/nextpage2"
        )
      )
    ).toMatchSnapshot();
  });
});
