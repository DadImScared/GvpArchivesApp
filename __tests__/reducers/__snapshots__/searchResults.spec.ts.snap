// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchResults reducer should add search 1`] = `
Object {
  "radha-book,movie": Object {
    "nextPage": "/nextpage",
    "results": Array [
      Object {
        "highlightedBody": "search body",
        "item_id": "1",
      },
    ],
    "suggestions": Array [],
    "timeToLive": 1558680261568,
  },
}
`;

exports[`searchResults reducer should update search 1`] = `
Object {
  "radha-book,movie": Object {
    "nextPage": "/nextpage2",
    "results": Array [
      Object {
        "highlightedBody": "search body",
        "item_id": "2",
      },
      Object {
        "highlightedBody": "search body",
        "item_id": "1",
      },
      Object {
        "highlightedBody": "search body",
        "item_id": "2",
      },
    ],
    "suggestions": Array [
      "suggestion",
    ],
    "timeToLive": 1558593861568,
  },
}
`;
